A55-stateful-session-affinity.md

```diff
@@ -0,0 +1,398 @@
+for maintaining session affinity. If the new subchannel goes into
+`TRANSIENT_FAILURE`, we just delegate to the child policy.
+
+Note that we create the `“override-host-experimental”` policy as the child of
+`“xds_cluster_impl”` only when the `override_host_status` config is present
+which is passed through
+[`common_lb_config.override_host_status`][or-host-status]. The
+`“cluster_resolver_experimental”` policy will receive this value in its CDS
+update and pass this config via its child (`“priority_experimental”`) to the
+`“xds_cluster_impl”` policy. Only when `override_host_status` is set it will
+create the `“override-host-experimental”` policy as its child and make one of
+the existing policies (`"wrr-locality-experimental"`, `"weighted_target"`
+or `"ring_hash_experimental"`) as the child policy of
+`“override-host-experimental”` as shown in the following diagram.
+
+![LB Policy Hierarchy Diagram](A55_graphics/lb-hierarchy.png)
+
+To propagate [`common_lb_config.override_host_status`][or-host-status] the
+following changes are required:
+
+* the [`XdsClient`][grpc-client-arch] to parse
+[`common_lb_config.override_host_status`][or-host-status] and copy the value
+in its CDS update to the watchers.
+
+* `cds_experimental` policy (which receives the CDS update) copies the value
+into `ClusterResolverConfig` for its child policy
+`cluster_resolver_experimental`. `ClusterResolverConfig` will be modified to
+contain this value.
+
+* `cluster_resolver_experimental` policy copies the value into
+`ClusterImplConfig` and passes that via its child policy
+`priority_experimental`.
+
+* `priority_experimental` policy extracts `ClusterImplConfig` from the
+config passed to it and uses that child config for creating its child
+`xds_cluster_impl` policy.
+
+* `xds_cluster_impl` extracts `override_host_status` from its config and uses
+that to conditionally create `override-host-experimental` as described above.
+
+
+
+## Rationale
+
+The stateful session affinity requirement is not satisfied by the
+[ring-hash LB policy][A42] because any time a backend host is removed or
+added, the hash mapping of 1/N client sessions gets affected (where N is the
+total number of backends) which would lead to dropped requests in some cases
+which is unacceptable.
+
+This feature avoids the problem by using HTTP cookies as described above.
+
+## Implementation
+
+The design will be implemented in gRPC C++, Java and Go - in that order.
+Wrapped languages should be able to use the feature provided they can do
+cookie management in their client code.
+
+Cookies are rarely used with gRPC so the implementations should also
+include a recipe for cookie management. This could involve illustrative
+examples using cookie jar implementations which are available in various
+languages such as Java, Go and Node.js. Wherever possible the
+implementations may also include a reference implementation of an interceptor
```

Comment:
Given that we expect the application to use a different cookie jar for each session, I'm not sure we can provide a common interceptor that will do that, since the interceptor does not know which RPCs belong to which sessions.  I think this is going to need to be done by the application code.

Reply:
While I agree it's better done by application code - I think a common cookie jar implementing Path would work.

Reply:
No, a single cookie jar implementing path will not work, because there will be many different sessions to the same path.  Only the application knows which RPCs are associated with which sessions; gRPC cannot determine that.

Reply:
Sessions are (under current proposal) not associated with RPCs/requests - but with DestinationRules and Service objects in k8s. The server will send a cookie for hostname and /service_name/ or /service_name/methodname - the client will store the cookie in the jar, and send it for requests with that service_name ( and methodname ). 

If session is done in application code, like HttpSession in java - the application code will decide the request should start a session so subsequent requests are sent to the same backend, will set a cookie with path /service_name[/method] - client will store it in the jar and future requests from that client will go to that backend.

Each service or method will have a different cookie - and the cookie jar code determines which cookie to send based on the http path.

Reply:
@costinm 

> Sessions are (under current proposal) not associated with RPCs/requests - but with DestinationRules and Service objects in k8s. The server will send a cookie for hostname and /service_name/ or /service_name/methodname - the client will store the cookie in the jar, and send it for requests with that service_name ( and methodname 

I don't think that's what we want. Only the application knows which RPCs are related and represent a separate session. So the cookie creation is done by the LB logic (occurring on the client side, not the server) and its the client app code that knows which cookie to use on which  RPC call. 

> If session is done in application code, like HttpSession in java - the application code will decide the request should start a session so subsequent requests are sent to the same backend, will set a cookie with path /service_name[/method] - client will store it in the jar and future requests from that client will go to that backend.

I am not too familiar with the HttpSession in Java but it seems that this is driven by the server side. In such case how would that work to properly affinitize through the LB stage that is occurring on the client. And such affinity needs to work even when the list backend host changes. The conventional constant-hashing LBs don't work well with that and hence only provide soft-affinity.

Reply:
> I'm not sure we can provide a common interceptor that will do that,...

The interceptor is per session (as mentioned a couple of lines below) so the user uses the interceptor only for a session.  For example in Java :

```
mySessionStub = TestServiceGrpc.newStub(channel).withInterceptors(new CookieManagingInterceptor());
// use mySessionStub for handling one stateful session.
``` 

Also this would be more of an example rather than a reference implementation.

Reply:
I didn't realize that interceptors were configured on a per-stub basis rather than a per-channel basis in Java.  Is that true in any other language?  I'm fairly sure it's not true in C++, but @yashykt can confirm.  For other languages, @dfawley for Go, @murgatroid99 for Node, @gnossen for Python.

In any case, this certainly would not work for C++, so even if it works for every single other language (which I doubt), it's clearly a language-specific thing, not an approach that can be used in all languages.

Reply:
Cookie created or managed by client code doesn't make any sense IMO, persistent sessions have been around for 20 years in java and I never heard of a client-specified cookie. It may be ok for client to explicitly set the IP for future requests - but that is not using a 'cookie', you can do it fully on client.

The persistent session needs to work across east/west gateways and other infra - in Istio for example we redirect requests trough multi-network gateways. And hopefully this mechanism can work for CloudRun and other services supporting session affinity, or for persistent sessions from internet clients. 

AFAIK Cookie protocol is defined as server setting the cookie, and client storing/passing them back.

Client only persistent sessions without middle boxes are already pretty simple - just Dial the IP of the backend and you have the desired behavior. There are plenty of ways to discover the IPs ( stateful sets in K8S, etc). No need for cookies or XDS to do this. What cookies allow is for the persistence to stick when middle boxes are in the way.

Reply:
@costinm The cookie here is created by the LB logic (which happens to be on the client side) not by the actual application client code. I would agree that normally it would be the service that creates cookies which map to specific state identity. We can take a similar approach but the challenge is that once such cookie is created all future requests bearing such cookie must be hard affinitized by the infrastructure to a specific backend. currently gRPC LB functionality doesn't support that. Envoy doesn't support that either and its model is to create the cookie as part of the LB logic (through a separate filter) and that cookie then keeps the LB decision state. I do see the point you are making about other intermediate nodes through which the flow must traverse in some topologies. The complete solution should be able to maintain such affinity through the entire path, not just the first hop. It seems that there is an implicit assumption that EDS always carries the final endpoint address (never an intermediary node), correct? 

> Client only persistent sessions without middle boxes are already pretty simple - just Dial the IP of the backend and you have the desired behavior. There are plenty of ways to discover the IPs ( stateful sets in K8S, etc). No need for cookies or XDS to do this. What cookies allow is for the persistence to stick when middle boxes are in the way.

The client gRPC application code doesn't have easy access to service discovery and EDS. Thats where service mesh and xDS come in. Therefore it makes no sense(in my mind) to try reimplement the logic that a service mesh (xDS) is already providing.

Reply:
In Java you can configure interceptors for the channel as well - so it allows both. In any case this implementation note only talks about providing an example (or preferably a reference implementation) and it doesn't have to be interceptor based if that doesn't work in some languages. I will add some text to that effect.

Reply:
>  I do see the point you are making about other intermediate nodes through which the flow must traverse in some topologies. The complete solution should be able to maintain such affinity through the entire path, not just the first hop.

The intermediate nodes could create their own cookies and add them to the RPC responses. As long as the client application copies those cookies to the following requests you have session affinity through the entire path, no?

Reply:
EDS certainly does not carry the final endpoint in Istio - that's how multi-network is implemented, all the endpoints in a different network receive the IP of the E-W gateway serving that network. Also I don't think we can define an API - in either Istio or gRPC - where only this strange and unprecedented Envoy behavior is supported. 

If the client LB is the only piece making the decision to keep requests going to same IP - why all this complexity with a cookie and all the other layers ? As I mentioned, client-side stickiness is currently  possible from client code ( Dial(IP) - if it fails Dial(service) and check peer IP for future requests). Persistent sessions using server-generated cookie and using server infra is also possible - clients can check MD for cookie and pass it. There is no change needed for either of this. 

I agree that it would be convenient for client to be able to pass the IP to the LB - like in Envoy, where a header or cookie
can be used - and avoid some of the complexity. But calling it 'client generated cookie' would be extremely confusing for anyone who used cookies, and supporting only this limited case - and not supporting the wast number of load balancers
 supporting 'normal' session persistent - is not ideal.  
  
At the minimum the client should store and pass back the cookie - if it can't use it directly - so all this infra can work.

Reply:
In normal session persistence, the cookie is an encrypted blob or used as a lookup key to a 'lookaside' service. 
The blob or lookaside service may return multiple results - for example a browser session may interact with multiple
backends using persistent sessions, and the final routing to the backend may happen at different stages ( global LB may forward to a regional LB first). In gRPC and envoy sidecar - we want to also support client directly routing to the sticky backend, but hopefully without breaking those other cases. 

It might be possible to do this using separate cookies - and letting user code deal with persistent sessions when intermediary LBs are in the path ( AFAIK this should already work ). So maybe all we need is to just check if server returns a cookie and use that instead ( assuming some LB or server is managing the persistence), otherwise just keep track the IP address. Or if the client code includes a 'cookie' that they tracked - use it. 

RFC6265: ".. by HTTP servers to store state (called cookies) at HTTP user agents", i.e. if it's the LB remembering it used an IP and not sending to other IPs as long as that is alive it probably is not a cookie.

Reply:
> The intermediate nodes could create their own cookies and add them to the RPC responses. As long as the client application copies those cookies to the following requests you have session affinity through the entire path, no?

I think that should work. 


> If the client LB is the only piece making the decision to keep requests going to same IP - why all this complexity with a cookie and all the other layers ? As I mentioned, client-side stickiness is currently possible from client code ( Dial(IP) - if it fails Dial(service) and check peer IP for future requests). Persistent sessions using server-generated cookie and using server infra is also possible - clients can check MD for cookie and pass it. There is no change needed for either of this.

As I mentioned earlier, ideally the server could generate a cookie that identifies the server state for a set of RPC requests. If we had a LB infra that could be configured to properly guarantee session affinity based of the server-side generated cookie that would be sufficient. But that's not the case. None of currently supported LB or routing modes in Envoy implement proper affinity (only soft affinity) based on an HTTP header or cookie. 

I don't think it should be the role of the client to have the logic to dial to specific endpoint IPs to achieve that affinity, including in cases where they are connecting through E-W gateways. In some topologies the clients are connecting to a VIP and would be seeing a single IP overall. When xDS comes into the picture the client has no knowledge or control of the endpoint being selected for each and every request.  

> I agree that it would be convenient for client to be able to pass the IP to the LB - like in Envoy, where a header or cookie can be used - and avoid some of the complexity. 

The client doesn't pass an IP to the LB, it treats the cookie as an opaque object. 

> But calling it 'client generated cookie' would be extremely confusing for anyone who used cookies, and supporting only this limited case - and not supporting the wast number of load balancers
> supporting 'normal' session persistent - is not ideal.


The cookie being discussed here is not 'client generated'. it is generated by the LB infra that happens to be part of the client gRPC core layer. 

> At the minimum the client should store and pass back the cookie - if it can't use it directly - so all this infra can work.

Thats exactly what is being assumed here. 


> It might be possible to do this using separate cookies - and letting user code deal with persistent sessions when intermediary LBs are in the path ( AFAIK this should already work ). So maybe all we need is to just check if server returns a cookie and use that instead ( assuming some LB or server is managing the persistence), otherwise just keep track the IP address. 

To make this work, there would need to be a [server-generated-cookie->serverIP] mapping state that needs to be maintained internally as part of the LB layer in gRPC and/or Envoy. This is not how it is currently implemented in Envoy.

Reply:
Agree with @va7ish 's last comment.

Reply:
@costinm It is definitely not feasible for clients to create a new channel directly to the IP address that they want the session to go to.  That approach would lose all sorts of important xDS-configured things (e.g., mTLS config, retry config, fault injection, etc), and it would lose the ability to fall back to a different endpoint in the event that the original one actually does go down.

That having been said, I do think that there is a general problem here, as pointed out by @ejona86 offline, which is that by using a different cookie jar for each session, the client application is making assumptions about what the xDS config is, which means that the xDS config cannot change things in ways that are incompatible with what the client is doing.  I think the example cited above (the traffic going through an intermediate proxy that may set its own cookie, which the client might wind up applying to just one session) is just one example of how this can be a problem.

I think we should consider whether there are ways to ameliorate this problem.  For example, instead of using cookies, which already have existing semantics that do not match what would need to happen here, what if we just used a configurable header?  That would still require explicit cooperation from the application, but the mechanism would not make any assumptions about existing semantics (e.g., cookies that are not intended to be tied to one particular session), so it may be less likely to break things.  The down-side is that it would be more work for the application to do this, because it could not just use an off-the-shelf cookie jar.  But the trade-off in flexibility may be worth that additional work.

Thoughts...?

Reply:
BTW, it looks like we aren't the only ones talking about possibly using a header instead of a cookie here.  I just noticed https://github.com/envoyproxy/envoy/pull/23145, which is proposing adding header-based support for stateful session affinity.

Reply:
@markdroth yes. Some of our customers are looking for this feature and hence that PR

Reply:
Do we need to include that feature in this design?

Reply:
No. It can wait for gRPC. We want primarily from Envoy side.

Reply:
I don't see any actionable comments in this thread. Resolve this conversation?

Reply:
We can always add the header-based affinity support to this gRFC as another PR as needed. I would like to finish this gRFC to include the DRAINING support (as discussed elsewhere) first.

Reply:
I am going to resolve this comment thread as there is nothing actionable AFAICS